### ISSUE TITLE
Pairs to consider during PredCls

### ISSUE THREAD

**Issue by nickgkan** [jz462/Large-Scale-VRD.pytorch#2]
Hello, thanks for the PyTorchh code release as well!

One question, during Predicate Classification, the objects' locations and labels are known, but how about the pairs tested?
I mean, do you examine all possible pairs or only the annotated? After briefly studying your code, I think you test only the annotated pairs, but correct me if I am wrong. I'm asking because the Neural Motifs paper seems to examine all possible pairs, I have not thoroughly tested their code though.

Thanks in advance!

**Comment by jz462** [comment]
Hi @nickgkan,

In fact I'm using all possible pairs by default, i.e., all pairs of objects during testing. What you said about using annotated pairs is true for the VG80K dataset from the Caffe2 code (https://github.com/facebookresearch/Large-Scale-VRD) but not for this repo. If you look at the "Evaluating Pre-trained Relationship Detection models" section in README, you'll see that by default the code tests with all pairs of objects produced by the object detector (the "SGDET" metric), and there are two options, "--use_gt_boxes" and "--use_gt_labels", that allow you to control whether to use the annotated boxes (the "SGCLS" metric) or both the annotated boxes and labels (the "PRDCLS" metric) as you like.

Hope this answer could help.

**Comment by naykun** [comment]
Hi @nickgkan & @jz462, after carefully checking, I do have the same question here. 
According to [model_builder_rel.py line 275-325](https://github.com/jz462/Large-Scale-VRD.pytorch/blob/7363b997f4cabae39759ce70b7f52c7a1d8d8d41/lib/modeling/model_builder_rel.py#L275), if the model is in test phase and roidb is not None, 'sbj_gt_boxes' and 'obj_gt_boxes' will be used as relation pairs directly. And in [test_engine_rel.py line 296](https://github.com/jz462/Large-Scale-VRD.pytorch/blob/7363b997f4cabae39759ce70b7f52c7a1d8d8d41/lib/core/test_engine_rel.py#L296), the dataset builder seems no difference with training dataset builder, since the "gt" argument is always True. Also, the evaluation annotation file only contains ground truth pairs without all possible boxes combination. As I think, that will leads to give the model gt sbj/obj pairs rather than gt boxes during SGCLS and PRDCLS task evaluation, which is not consistent with the origin task definition. And indeed, it will lead to higher performance.
I'm wondering if I have any misunderstandings here. Could you please provide a more detailed explanation about how your code works on testing SGCLS and PRDCLS? The above answer is still confusing to me.
Thanks very much!

**Comment by sigeek** [comment]
I think this is related to the issue mentioned [here](https://github.com/NVIDIA/ContrastiveLosses4VRD/issues/20). They evaluate SGCls and PredCls given triplets, not the complete set of objects in the image.
